zap/src/main/java/org/zaproxy/zap/session/ScriptBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,752 @@
+                                    SCRIPT_ICON_SESSION,
+                                    false,
+                                    new String[] {ScriptType.CAPABILITY_APPEND}));
+        }
+    }
+
+    @Override
+    public SessionManagementMethod loadMethodFromSession(Session session, int contextId)
+            throws DatabaseException {
+        ScriptBasedSessionManagementMethod method = new ScriptBasedSessionManagementMethod();
+
+        List<String> scripts =
+                session.getContextDataStrings(
+                        contextId, RecordContext.TYPE_SESSION_MANAGEMENT_FIELD_1);
+        if (scripts.size() > 0) {
+            // Load the script and make sure it still exists and still follows the required
+            // interface
+            this.loadMethod(
+                    method,
+                    scripts.get(0),
+                    session.getContextDataStrings(
+                            contextId, RecordContext.TYPE_SESSION_MANAGEMENT_FIELD_2));
+        }
+        return method;
+    }
+
+    @Override
+    public void exportData(Configuration config, SessionManagementMethod sessionMethod) {
+        if (!(sessionMethod instanceof ScriptBasedSessionManagementMethod)) {
+            throw new UnsupportedSessionManagementMethodException(
+                    "Script based session management type only supports: "
+                            + ScriptBasedSessionManagementMethod.class.getName());
+        }
+        ScriptBasedSessionManagementMethod method =
+                (ScriptBasedSessionManagementMethod) sessionMethod;
+        if (method.script != null) {
+            config.setProperty(CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_NAME, method.script.getName());
+        }
+        config.setProperty(
+                CONTEXT_CONFIG_SESSION_MGMT_SCRIPT_PARAMS,
+                EncodingUtils.mapToString(method.paramValues));
+    }
+
+    @Override
+    public void persistMethodToSession(
+            Session session, int contextId, SessionManagementMethod method)
+            throws DatabaseException {
+
+        if (!(method instanceof ScriptBasedSessionManagementMethod)) {
+            throw new UnsupportedSessionManagementMethodException(
+                    "Script based session management type only supports: "
+                            + ScriptBasedSessionManagementMethod.class.getName());
+        }
+
+        ScriptBasedSessionManagementMethod scriptMethod =
+                (ScriptBasedSessionManagementMethod) method;
+        if (scriptMethod.script != null) {
+            session.setContextData(
+                    contextId,
+                    RecordContext.TYPE_SESSION_MANAGEMENT_FIELD_1,
+                    scriptMethod.script.getName());
+        }
+        session.setContextData(
+                contextId,
+                RecordContext.TYPE_SESSION_MANAGEMENT_FIELD_2,
+                EncodingUtils.mapToString(scriptMethod.paramValues));
```

Comment:
This will also need a null check, `paramValues` might be null, or change it to be never null?